[HLSL][docs] Document hlsl.h in the HLSL docs (#84081)
llvm/llvm-project: clang/docs/HLSL/HLSLSupport.rst

with other GPU graphics targets like SPIR-V and possibly other GPU and even CPU
targets.

hlsl.h
------

> HLSL has an extensive library of functionality. This is similar to OpenCL and

Reviewer: ```suggestion
HLSL has a library that supports extensive set of functionalities. This is similar to OpenCL and
```

CUDA. The implementation approach for the HLSL library functionality draws from

> patterns in use by OpenCL and other Clang resource headers.

Reviewer 2: Would it be possible to have links to these similar headers for reference?

Similar to OpenCL, the HLSL library functionality is implicitly declared in
translation units without needing to include a header to provide declarations.
In Clang this is handled by making ``hlsl.h`` an implicitly included header
distributed as part of the Clang resource directory.

Similar to OpenCL, HLSL's implicit header will explicitly declare all overloads,

> and each overload will map to a corresponding ``__builtin_*`` compiler intrinsic

Reviewer 3: __builtin_* --> __builtin* 

seems to be how you do it elsewhere.

that is handled in ClangCodeGen. CUDA uses a similar pattern although many CUDA
functions have full definitions in the included headers which in turn call

> corresponding ``__builtin*`` compiler intrinsics. By not having bodies HLSL

Reviewer 4: Is it possible to define the body for some hlsl intrinsic functions like https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-d3dcolortoubyte4 which doesn't map to either DXIL or SPIRV?

Author: Possible, sure, but there is a performance impact do doing that because the function will need to be parsed, generated and inlined.

avoids the need for the inliner to clean up and inline large numbers of small
library functions.

HLSL's implicit headers also define some of HLSL's typedefs. This is consistent
with how the AVX vector header is implemented.

Concerns have been expressed that this approach may result in slower compile
times than the approach DXC uses where library functions are treated more like
Clang ``__builtin*`` intrinsics. No real world use cases have been identified
where parsing is a significant compile-time overhead, but the HLSL implicit
headers can be compiled into a module for performance if needed.

> Further, by treating these functions as functions rather than ``__builtin*``

Reviewer: ```suggestion
Further, by treating these as traditional language functions rather than as ``__builtin*`` 
```

> compiler intrinsics the language behaviors are more consistent and aligned with

Reviewer: ```suggestion
compiler intrinsics, the language behaviors are more consistent and aligned with
```

with other GPU graphics targets like SPIR-V and possibly other GPU and even CPU
targets.

hlsl.h
------

HLSL has a library of standalone functions. This is similar to OpenCL and CUDA,
and is analogous to C's standard library. The implementation approach for the
HLSL library functionality draws from patterns in use by OpenCL and other Clang
resource headers. All of the clang resource headers are part of the
``ClangHeaders`` component found in the source tree under
`clang/lib/Headers <https://github.com/llvm/llvm-project/tree/main/clang/lib/Headers>`_.

> .. note::

Author: Comment for reviewers: GitHub's preview doesn't properly render the `note` block. An example of this in HTML can be found here: https://clang.llvm.org/docs/LanguageExtensions.html#has-builtin
